### ISSUE TITLE
Sandi Metz' Rules for Developers

### ISSUE THREAD

**Issue by worace** [turingschool/intermission-assignments#35]
Discuss Sandi Metz's Rules for Developers here. Which of Sandi's rules do you feel like might be the hardest to follow—why?

**Comment by roseak** [comment]
Only instantiating one object in the controller would definitely have been tough with my personal project.  Also, breaking up if/else statements so that there are only two conditions each time that are one line each would need some finagling.  I can see how the rules would lead to really clean and thought out code though.

**Comment by jbrr** [comment]
The 5 line method rule is hard to follow, but I think the one I would have the most trouble with is the last one, controllers can only instantiate one object. I remember that was feedback that we got on the Pivot, is that we should only have one instance variable per controller action. That made sense, that became a lot harder with the personal project. We used AngularJS, and sort of the equivalent of `@foo` is `$scope.foo`, and we had at least 9 different `$scope`'s in our controller. Everything I saw in the documentation and different examples led me to believe that this is just how things are done in Angular, but I also realize that Sandi Metz' rules aren't just Rails specific. I'm sure our code quality isn't good, so I'd be curious about how to fix it in our project specifically, and also in general in other frameworks besides Rails.

**Comment by stevekinney** [comment]
@roseak and @jbrr Regarding conditionals in methods, Sandi [did an amazing talk at RailsConf this year](https://www.youtube.com/watch?v=OMPfEXIlTVE) about conditionals in object-oriented programming languages.

**Comment by rasensio1** [comment]
Hardest to follow - "no elsif". What happens when there are three, equally viable paths?

**Comment by stevekinney** [comment]
@rasensio1 You will never, ever see me write an `else if` in JavaScript or an `eslif` in Ruby. For me personally, that's a code smell and a sign that I need to consider refactoring or contemplating my design.

**Comment by rasensio1** [comment]
@stevekinney will watch Sandi's talk. I realize that my comment might be just be 'philosophical noodling' because I can't think of an instance where this is the case.

**Comment by jxandery** [comment]
"Controllers can instantiate only one object - is probably the hardest rule to follow. Often, we need more than one type of thing on a page." 

It's not so much that we need more than one type of thing on a page. Rather, it's when you're just trying to get something working, worrying about the view only having one instance variable too early can lead you down a rabbit hole. That rabbit hole might actually lead you further from a solution. Function first, then form.

**Comment by mbburch** [comment]
I'm thinking in line with those of you writing that controllers only instantiating one object is probably the trickiest for me. I'll also have to check out that podcast, @stevekinney! I can see trying to stay away from elsif/ else if, but can't imagine never using it. On the note of short methods though, I did like her point of private methods being good documentation.

**Comment by 7hoenix** [comment]
5 line methods.

And not even because of the "elsif" problem... I just naturally tend to want to solve something in one long method and then I conveniently "forget" to go back and refactor it later.

**Comment by plato721** [comment]
4 parameters/method.. the kicker is the hash options being counted separately. And the author's team had the same problem I would: forms. My personal project had like 17 things that could be configured by the user in a search.

**Comment by George-Hudson** [comment]
I agree with Matt. "Hash options are parameters": There often seems to be occasions where an argument might contain a hash with more than 4 key value pairs. This seems unnecessary. I can see occasions where this rule would create additional complexity with no benefit. It's a good thing each rule is allowed to be broken if a pair agrees!

**Comment by selfup** [comment]
@jphoenix86 I agree with the 5 line methods being a bit restrictive sometimes.

I have like 2 methods in my personal project that pretty much have to be more than 5 lines long. Just because of how things are called/loaded.

I will try my best to use this for JS, but it seems to a good set of rules, that unfortunately will be broken every now and then.

**Comment by stevekinney** [comment]
The trick here is that there is no limits on the number of methods you're allowed to have. It's basically a rule that encourages you to think about breaking down a large method into component parts.

In my time as an instructor, I've never seen a case where anyone (student or instructor) had too many single-purpose classes with too many small, succinct methods.
